R/ResampleInstances.R

```diff
@@ -8,10 +8,52 @@ instantiateResampleInstance.HoldoutDesc = function(desc, size, task = NULL) {
 }
 
 instantiateResampleInstance.CVDesc = function(desc, size, task = NULL) {
-  if (desc$iters > size)
-    stopf("Cannot use more folds (%i) than size (%i)!", desc$iters, size)
-  test.inds = chunk(seq_len(size), shuffle = TRUE, n.chunks = desc$iters)
-  makeResampleInstanceInternal(desc, size, test.inds = test.inds)
+
+  # Random sampling CV
+  if (!desc$fixed) {
+    if (desc$iters > size) {
+      stopf("Cannot use more folds (%i) than size (%i)!", desc$iters, size)
+    }
+    test.inds = chunk(seq_len(size), shuffle = TRUE, n.chunks = desc$iters)
+    makeResampleInstanceInternal(desc, size, test.inds = test.inds)
+  } else {
+
+    # CV with only predefined indices ("grouping")
+
+    if(is.null(task$blocking)) {
+      stopf("To use blocking in resampling, you need to pass a factor variable when creating the task!")
+    }
+
+    if (desc$iters != length(levels(task$blocking))) {
+      desc$iters = length(levels(task$blocking))
+    }
+    levs = levels(task$blocking)
+    n_levels = length(levels(task$blocking))
+
+    # Why do we need the helper desc? If we would call 'instantiateResampleInstance()' here,
+    # we would call the function within itself and will receive an 'error-c-stack-usage-is-too-close-to-the-limit' error
+    # So we simply change the class name to mimic a new function..
+    attr(desc, "class")[1] = "CVHelperDesc"
+    # create fake ResampleInstance
+    inst = instantiateResampleInstance(desc, n_levels, task)
+    attr(desc, "class")[1] = "CVDesc"
+
+    # now exchange block indices with indices of elements of this block and shuffle
+    test.inds = lapply(inst$test.inds, function(i) which(task$blocking %in% levs[i]))
+
+    # Nested resampling: We need to create a list with length(levels) first.
+    # Then one fold will be length(0) because we are missing one factor level because we are in the inner level
```

Comment:
What happens if the number of outer folds is less than the number of levels (or a simple train/test split) and more than one factor level is missing?

Reply:
In https://github.com/mlr-org/mlr/blob/192355ca3decbfe8d27d6a658a3072c7e2ac6442/R/ResampleInstances.R#L27-L28 we check that the number of folds is always = number of levels.

Reply:
Ah ok. Could you add information on what the number of levels was and what it was set to in the warning please?

Reply:
The problem we have here is the following:

* The warning is triggered for both inner and outer level
* Determining the level (inner/outer) is tricky
* The inner level resets to the factor levels of the outer level first and is then further adjusted. So we actually get a false positive for the inner level even if its set correctly (e.g. outer = 5, inner = 4). 

Even if this doesn't sound logical, I would even vote for the complete removal of the warning. Users who set `fixed = T` usually know what they want and what they need to set. 
Its contra-productive if the warning is raised even if the specification is correct (e.g. outer = 5, inner = 4). 

Not sure if this thinking if easy to follow here. Let me know if I should explain it again in more detail.

I would propose to mention the adjustment in the tutorial (and in the help page?).

Reply:
But at some point you can figure out whether there are enough levels for the folds, right? So there shouldn't need to be any false positives?

Reply:
Further down in the function I have the final number of levels. But then I have no information if the level count was adjusted or not. And even if I would add a flag, I am missing the original level count since `desc$iters` is reassigned. 

I think the effort of implementing a robust warning for both inner and outer is not worth the effort.
I would prefer to note it in the help page (in details) and in the tutorial.

Something like "Setting iters with fixed = T has no effect. iters will be set to length(blocking.levels) in the outer and length(blocking.levels) - 1 in the inner level".

Reply:
Thanks for discussing this Lars! It's also not a solution I am completely happy with and a problem I thought a lot about. 

I'll do the required changes soon, including the tutorial page. Finally getting this done.